x/p2p/client.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	ErrAppRequestFailed = errors.New("app request failed")
+	ErrRequestPending   = errors.New("request pending")
+	ErrNoPeers          = errors.New("no peers")
+)
+
+// AppResponseCallback is called upon receiving an AppResponse for an AppRequest
+// issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type AppResponseCallback func(
+	nodeID ids.NodeID,
+	responseBytes []byte,
+	err error,
+)
+
+// CrossChainAppResponseCallback is called upon receiving an
+// CrossChainAppResponse for a CrossChainAppRequest issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type CrossChainAppResponseCallback func(
+	chainID ids.ID,
+	responseBytes []byte,
+	err error,
+)
+
+type Client struct {
+	handlerID uint8
+	router    *Router
+	sender    common.AppSender
+}
+
+// AppRequestAny issues an AppRequest to an arbitrary node decided by Client.
+// If a specific node needs to be requested, use AppRequest instead.
+// See AppRequest for more docs.
+func (c *Client) AppRequestAny(
+	ctx context.Context,
+	appRequestBytes []byte,
+	onResponse AppResponseCallback,
+) error {
+	c.router.lock.RLock()
+	peers := maps.Keys(c.router.peers)
+	c.router.lock.RUnlock()
+
+	if len(peers) == 0 {
+		return ErrNoPeers
+	}
+
+	s := sampler.NewUniform()
+	s.Initialize(uint64(len(peers)))
+	i, err := s.Next()
+	if err != nil {
+		return err
+	}
+	nodeIDs := set.Set[ids.NodeID]{peers[i]: struct{}{}}
+	return c.AppRequest(ctx, nodeIDs, appRequestBytes, onResponse)
+}
+
+// AppRequest issues an arbitrary request to a node.
+// [onResponse] is invoked upon an error or a response.
+func (c *Client) AppRequest(
+	ctx context.Context,
+	nodeIDs set.Set[ids.NodeID],
+	appRequestBytes []byte,
+	onResponse AppResponseCallback,
+) error {
+	c.router.lock.Lock()
+	defer c.router.lock.Unlock()
+
+	appRequestBytes = c.prefixMessage(appRequestBytes)
+	for nodeID := range nodeIDs {
+		requestID := c.router.requestID
+		if _, ok := c.router.pendingAppRequests[requestID]; ok {
+			return fmt.Errorf(
+				"failed to issue request with request id %d: %w",
+				requestID,
+				ErrRequestPending,
+			)
+		}
+
+		if err := c.sender.SendAppRequest(
+			ctx,
+			set.Set[ids.NodeID]{nodeID: struct{}{}},
+			requestID,
+			appRequestBytes,
+		); err != nil {
+			return err
+		}
+
+		c.router.pendingAppRequests[requestID] = onResponse
+		c.router.requestID++
```

Comment:
Instead of holding the lock through the entire loop would it make sense to add this logic to a private method that releases the lock on each iteration?